[docs] Standardize all MUI Core "Usage" pages (#34183)
mui/material-ui: docs/data/joy/getting-started/usage/usage.md

# Usage

<p class="description">Learn the basics of using Joy UI components.</p>

## Getting started

The following code snippet demonstrates a simple app that uses the Joy UI [Button](/joy-ui/react-button/) component:

```jsx
import * as React from 'react';
import { CssVarsProvider } from '@mui/joy/styles';
import Button from '@mui/joy/Button';

export default function App() {
  return (
    <CssVarsProvider>
      <Button>Hello World</Button>
    </CssVarsProvider>
  );
}
```

### CssVarsProvider

In the example above, you can see that the Button component is nested within `<CssVarsProvider />`.

> This is not required, but it unlocks a whole host of customization options powered by CSS variables.

Reviewer: @samuelsycamore All Joy UI components require `CssVarsProvider` at the root to provide CSS theme variables.

Author: Ah my mistake, I'm not sure where I got the idea that it was optional.

See [Using CSS variables](/joy-ui/customization/using-css-variables/) for more details.

## Shared props

> Joy UI components are self-supporting and fully functional in isolation.

Reviewer: I do not quite understand what this line means.

Reviewer 2: If we keep this sentence on, I'm feeling like it would be best put it right below the first Getting started heading?! I get that at this current spot, it's a hook for the shared props stuff but it seems something we may want to share right away?

Author: Hmm, it's something that I original found in the Material UI Usage page, and copied it over to Base and then Joy. As I understand it, the idea is that every individual component can be used by itself in an app—it doesn't require the use of any other components to work correctly. In order words, you don't have to build a strictly Joy UI-only app in order to use a Joy component as a one-off in a project.

Reviewer 3: this is accurate, it's also about the idea that you can add this inside an existing Bootstrap application, and you won't break everything nor will it break your component because they use box-sizing: box-content and not border-content.

Each component has its own unique API, but all components accept the following shared props:

> ### components

Reviewer: Although, this is a very useful section I feel that it should be on a different page that explains "slot overrides".

To me, the usage page should be simple as possible to start using components.

I believe that the slot overrides are not common use cases and they should be in an advanced guide.

Reviewer 2: Agreed. We should probably mention somewhere that everything as far as these shared props goes are exactly the same as Base, since it is using Base components under the hood anyway.

Author: That's interesting feedback. MUI Base has a 4th page in the "Getting started" sequence after Usage called [Customization](https://mui.com/base/getting-started/customization/). Maybe this page would be a better place to document these more advanced props?

Reviewer: Agreed.

Author: Yeah I think that makes sense. In that case, I'll remove that info from these pages in this PR, and create a new PR to standardize the "Customization" pages and add one for Joy. Does that sound good?

Reviewer 2: Sounds good!

# Usage

<p class="description">Learn the basics of using Joy UI components in two quick steps.</p>
<p class="description">Learn the basics of using Joy UI components.</p>

## Getting started
## Quickstart

### Set up the `CssVarsProvider` component

Go to your `App.js` file and replace it with the code snippet below.

You should see the text `Hello from Joy` being rendered on your browser.
The following code snippet demonstrates a simple app that uses the Joy UI [Button](/joy-ui/react-button/) component:

```jsx
import * as React from 'react';
import { CssVarsProvider } from '@mui/joy/styles';
import Button from '@mui/joy/Button';

function App() {
  return <CssVarsProvider>Hello from Joy</CssVarsProvider>;
}

export default App;
```

### Render your first component

To render any Joy UI component, make sure you place them inside the `CssVarProvider` element.

```diff
import { CssVarsProvider } from '@mui/joy/styles';
+ import Button from '@mui/joy/Button';

function App() {
export default function App() {
  return (
    <CssVarsProvider>
+     <Button>Joy UI</Button>
      <Button>Hello World</Button>
    </CssVarsProvider>
  );
}

> export default App;

Reviewer 3: I think that it would be great to have a working iframe codesandbox to start with like for Material UI. It can make playing with the API so much easier.

Reviewer: I think the idea.

Author: Good call!

> ### CssVarsProvider

Reviewer 3: What about the `### Responsive meta tag`?

# Usage

<p class="description">Learn the basics of using Joy UI components.</p>

## Quickstart

The following code snippet demonstrates a simple app that uses the Joy UI [Button](/joy-ui/react-button/) component:

```jsx
import * as React from 'react';
import { CssVarsProvider } from '@mui/joy/styles';
import Button from '@mui/joy/Button';

export default function App() {
  return (
    <CssVarsProvider>
      <Button variant="solid">Hello World</Button>
    </CssVarsProvider>
  );
}
```

You can play around with this code in the interactive Code Sandbox demo below.
Try changing the `variant` on the Button to `soft` to see how the style changes:

> {{"demo": "Usage.js", "hideToolbar": true, "bg": true}}

Reviewer 3: From what I understand, we also need to load the default font for Joy UI.

Reviewer 3: Don't we also miss it from the codesandbox export of the demos? e.g. https://codesandbox.io/s/c50w6x?file=/public/index.html. I don't personally have Public Sans.

Author: Following up, my understanding is that Public Sans is bundled directly into `@mui/joy` so there's no need to load it in. Is that correct @siriwatknp?

Reviewer 3: @samuelsycamore fonts can't be bundled directly, they have to be installed somehow. For example

- default Joy UI: https://codesandbox.io/s/misty-sky-n5ct34?file=/public/index.html
- with Public Sans installed: https://codesandbox.io/s/cocky-herschel-3gk5k5?file=/public/index.html

To compare with https://mui.com/joy-ui/react-typography/#levels. I feel like we might not actually want to use Public Sans, or June and Danilo have the font already installed on their OS so they were blind to the problem.

Author: @oliviertassinari that makes sense. I'd like to get this PR merged, and I think we should continue the conversation about the font somewhere else. How should we proceed to close this out? We could:

1. leave the demo as is and don't mention the font in the doc (for now, assuming it's going to change very soon)
2. import Public Sans in the demo via Google Fonts, and add info about installing the font to Joy UI's Installation page as we've done for Roboto with Material UI

The first is the fastest and simplest option, but I'd go with the second option in the interest of making these docs fully functional now, whether or not Public Sans or its installation change in the future. We'll also want to update the [Joy UI templates](https://mui.com/joy-ui/getting-started/templates/) to replace [the MD/Roboto stuff](https://codesandbox.io/s/rjnzeb?file=/public/index.html) with Public Sans.

cc: @siriwatknp @danilo-leal

Reviewer 2: I'd be down to go with the first option there first, taking advantage of Joy's still in-alpha state to iterate on the typeface but would definitely understand if you understand is better to document it already and then, later on, we would just switch over to the new/different typeface if that happens, and whatever it ends up being!

Reviewer 3: @danilo-leal If we go with option 1, should we remove Public Sans from the source? This way, we will have a more predictable font experience. I mean, this way, the font displayed won't change based on whether I have "Public Sans" installed on my OS or not.

Author: Personally, I think I would be kind of frustrated if I was trying out Joy UI for a project and couldn't find any info about fonts (aside from what's on the Typography page)—especially if I'm coming from Material UI, where that question is clearly answered in the docs. Even if it's still in alpha and could potentially change, I'd rather have a solution that works now.

Reviewer 2: Alright, that definitely makes sense. Let's add documentation around the font now, keep it in the source, and then if that changes at any time, we just update everything to the new typeface. Does that sound good?

### CssVarsProvider

In the example above, you can see that the Button component is nested within `<CssVarsProvider />`.

> This provider unlocks a whole host of customization options powered by CSS variables.

Reviewer 3: ```suggestion
**This provider is required**, it unlocks a whole host of customization options powered by CSS variables.
```

We could almost add an alter callout. In Material UI, we used to have a required provider like this, but MANY developers were struggling with it, so we gave up on the requirement when moving to v1. The space might have changed since then, developers are now probably more educated on how to handle a global provider.

In any case, @siriwatknp I think that this issue is still relevant for us: #5738. In https://codesandbox.io/s/joy-ui-forked-282rws?file=/index.tsx, it doesn't work, but you have no clue why in the console.

Reviewer: > In any case, @siriwatknp I think that this issue is still relevant for us: https://github.com/mui/material-ui/issues/5738. In https://codesandbox.io/s/joy-ui-forked-282rws?file=/index.tsx, it doesn't work, but as a developer, you have no clue about how to fix it in the console.

I'd say that it is the DX issue that should be improved.

Reviewer: But I agree with @oliviertassinari 's suggested changes.

Author: Good call @oliviertassinari - I think you're right that this deserves a callout, especially since you don't get any errors when you omit the provider.

See [Using CSS variables](/joy-ui/customization/using-css-variables/) for more details.

> 

Reviewer 3: I would have expected a `## Globals` on this page. From what I understand, we need to document the meta, the font and the CssBaseline. Actually, I see no CssBaseline in Joy UI: https://codesandbox.io/s/joy-ui-forked-j8qp8i?file=/index.tsx. How do we solve the problems that CssBaseline solve?

Reviewer: Yeah, we should have CssBaseline in Joy UI as well. Added to the list.

Author: I skipped the "Globals" header here since it only contained one item. Would `<CssVarsProvider>` also be considered a "global"?

# Usage

> <p class="description">Learn the basics of using Joy UI components.</p>

Reviewer 3: To match Material UI and MUI Base?

```suggestion
<p class="description">Learn the basics of working with Joy UI components.</p>
```

## Quickstart

The following code snippet demonstrates a simple app that uses the Joy UI [Button](/joy-ui/react-button/) component:

```jsx
import * as React from 'react';
import { CssVarsProvider } from '@mui/joy/styles';
import Button from '@mui/joy/Button';

> export default function App() {

Reviewer 3: To match Material UI and MUI Base?

```suggestion
export default function MyApp() {
```
